src/dbnode/persist/fs/seek_manager.go

```diff
@@ -286,15 +343,188 @@ func (m *seekerManager) Return(shard uint32, start time.Time, seeker ConcurrentD
+//      running the UpdateOpenlease() function that all inactive seekers have been returned and closed at
+//      which point the function will return sucessfully.
 func (m *seekerManager) UpdateOpenLease(
 	descriptor block.LeaseDescriptor,
 	state block.LeaseState,
 ) (block.UpdateOpenLeaseResult, error) {
-	// TODO(rartoul): This is a no-op for now until the logic for swapping out seekers is written.
-	// Also make sure that this function is a proper no-op if the SeekerManager state has been
-	// been switched to closed.
-	return block.NoOpenLease, nil
+	m.Lock()
+	if m.status != seekerManagerOpen {
+		m.Unlock()
+		return 0, errUpdateOpenLeaseSeekerManagerNotOpen
+	}
+
+	if m.isUpdatingLease {
+		// This guard is a little overly aggressive. In practice, the algorithm remains correct even in the presence
+		// of concurrent UpdateOpenLease() calls as long as they are for different shard/blockStart combinations.
+		// However, the calling code currently has no need to call this method concurrently at all so use the
+		// simpler check for now.
+		m.Unlock()
+		return 0, errConcurrentUpdateOpenLeaseNotAllowed
+	}
+
+	if !m.namespace.Equal(descriptor.Namespace) {
+		m.Unlock()
+		return block.NoOpenLease, nil
+	}
+	m.isUpdatingLease = true
+	m.Unlock()
+	defer func() {
+		m.Lock()
+		m.isUpdatingLease = false
+		m.Unlock()
+	}()
+
+	// First open a new seeker outside the context of any locks.
+	seeker, err := m.newOpenSeekerFn(
+		descriptor.Shard, descriptor.BlockStart, state.Volume)
+	if err != nil {
+		return 0, err
+	}
+
+	newActiveSeekers, err := m.seekersAndBloomFromSeeker(seeker, state.Volume)
+	if err != nil {
+		// Don't need to worry about closing / leaking seeker here because
+		// seekersAndBloomFromSeeker will have closed it already if there
+		// were any errors.
+		return 0, err
+	}
+
+	var (
+		byTime                = m.seekersByTime(descriptor.Shard)
+		blockStartNano        = xtime.ToUnixNano(descriptor.BlockStart)
+		updateOpenLeaseResult = block.NoOpenLease
+
+		seekers rotatableSeekers
+		ok      bool
+	)
+	// It's possible that another goroutine is currently trying to open seekers for this blockStart. If so, this
+	// goroutine will need to wait for the other goroutine to finish before proceeding. The check is performed in
+	// a loop because each iteration relinquishes the lock temporarily. Once the lock is reacquired the same
+	// conditions need to be checked again until this Goroutine finds that either:
+	//
+	//   a) Seekers are already present for this blockStart in which case it can proceed to hot swap them.
+	// or
+	//   b) Seeks are not present for this blockStart and no other goroutines are currently trying to open them,
+	//      in which case this goroutine can just set the new seekers and be done.
+	for {
+		byTime.Lock()
+		seekers, ok = byTime.seekers[blockStartNano]
+
+		if !ok || seekers.active.wg == nil {
+			// Exit the loop still holding the lock.
+			break
```

Comment:
This definitely seems somewhat dangerous to break the loop and expect the following code to always `byTime.Unlock()` without a defer being used somewhere.

Can break this into sub functions to guarantee the locking lock and unlock semantics? 

I don't see any errors with this code, but someone making a change could definitely introduce a bug that could potentially slip past tests unless it's more straight forward to reason about.

Reply:
Yeah I just refactored it so the locking and such is more straightforward